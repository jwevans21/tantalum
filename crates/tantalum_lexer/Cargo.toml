[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
