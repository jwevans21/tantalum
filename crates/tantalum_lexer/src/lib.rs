//! # Tantalum Lexer
//!
//! This crate provides a lexer for the Tantalum language.

use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxToken};

/// A lexer for the Tantalum language.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Lexer<'file> {
    /// The source file that the lexer is currently lexing.
    file: &'file SourceFile,
    /// The current location in the source file.
    location: SourceLocation,
}

impl<'file> Lexer<'file> {
    /// Create a new lexer from a source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let lexer = Lexer::new(file);
    /// ```
    #[must_use]
    pub fn new(file: &'file SourceFile) -> Self {
        Self {
            file,
            location: file.location_start(),
        }
    }

    /// Get the source file that the lexer is currently lexing.
    #[must_use]
    pub fn source(&self) -> &'file SourceFile {
        self.file
    }

    /// Get the current location in the source file.
    #[must_use]
    pub fn location(&self) -> SourceLocation {
        self.location
    }

    /// Get the character at a location in the source file.
    fn at_location(&self, location: SourceLocation) -> Option<char> {
        self.file.read_char(location)
    }

    /// Extract the next token from the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.next_token().expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    /// ```
    #[must_use]
    pub fn next_token(&mut self) -> Option<SyntaxToken> {
        let start = self.location;

        let (kind, end) = self.characters_to_token(start)?;

        self.location = end;

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Peek at the next token in the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.peek_token(1).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    ///
    /// let token = lexer.peek_token(2).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Whitespace);
    ///
    /// let token = lexer.peek_token(3).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Identifier);
    /// ```
    #[must_use]
    pub fn peek_token(&self, mut n: usize) -> Option<SyntaxToken> {
        let mut start = self.location;

        let (kind, end) = loop {
            n -= 1;
            let (kind, end) = self.characters_to_token(start)?;

            if n == 0 {
                break (kind, end);
            }

            start = end;
        };

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Convert a character to a [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    fn characters_to_token(&self, start: SourceLocation) -> Option<(SyntaxKind, SourceLocation)> {
        let mut current = start;

        let c = self.at_location(current)?;
        current = current.next_by(c.len_utf8());

        let token = match c {
            '(' => SyntaxKind::LParen,
            ')' => SyntaxKind::RParen,
            '{' => SyntaxKind::LBrace,
            '}' => SyntaxKind::RBrace,
            '[' => SyntaxKind::LBracket,
            ']' => SyntaxKind::RBracket,
            '<' => SyntaxKind::LAngle,
            '>' => SyntaxKind::RAngle,
            ',' => SyntaxKind::Comma,
            ':' => SyntaxKind::Colon,
            ';' => SyntaxKind::Semicolon,
            '.' => SyntaxKind::Dot,
            '&' => SyntaxKind::Ampersand,
            '|' => SyntaxKind::Pipe,
            '^' => SyntaxKind::Caret,
            '+' => SyntaxKind::Plus,
            '-' => SyntaxKind::Minus,
            '*' => SyntaxKind::Star,
            '/' => SyntaxKind::Slash,
            '\\' => SyntaxKind::Backslash,
            '%' => SyntaxKind::Percent,
            '=' => SyntaxKind::Equals,
            '!' => SyntaxKind::Bang,
            '~' => SyntaxKind::Tilde,
            '?' => SyntaxKind::Question,
            '@' => SyntaxKind::At,
            '0'..='9' => SyntaxKind::Digit,
            '_' | 'a'..='z' | 'A'..='Z' => {
                while let Some(c) = self.at_location(current) {
                    match c {
                        '_' | 'a'..='z' | 'A'..='Z' | '0'..='9' => {
                            current = current.next_by(c.len_utf8());
                        }
                        _ => break,
                    }
                }

                match self.file.read_span(SourceSpan::new(
                    self.file.id(),
                    self.location.offset(),
                    current.offset(),
                ))? {
                    "fn" => SyntaxKind::Fn,
                    "let" => SyntaxKind::Let,
                    "return" => SyntaxKind::Return,
                    "if" => SyntaxKind::If,
                    "else" => SyntaxKind::Else,
                    "true" => SyntaxKind::True,
                    "false" => SyntaxKind::False,
                    _ => SyntaxKind::Identifier,
                }
            }
            c if c.is_whitespace() => SyntaxKind::Whitespace,
            _ => SyntaxKind::Other,
        };

        Some((token, current))
    }
}

#[test]
fn keywords() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "fn let return if else true false".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 7);

    assert_eq!(tokens[0].kind(), SyntaxKind::Fn);
    assert_eq!(tokens[1].kind(), SyntaxKind::Let);
    assert_eq!(tokens[2].kind(), SyntaxKind::Return);
    assert_eq!(tokens[3].kind(), SyntaxKind::If);
    assert_eq!(tokens[4].kind(), SyntaxKind::Else);
    assert_eq!(tokens[5].kind(), SyntaxKind::True);
    assert_eq!(tokens[6].kind(), SyntaxKind::False);
}

#[test]
fn identifiers() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "a A _ a1 A1 _1 _a _A __AA01".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 9);

    assert_eq!(tokens[0].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[1].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[2].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[3].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[4].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[5].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[6].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[7].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[8].kind(), SyntaxKind::Identifier);
}
