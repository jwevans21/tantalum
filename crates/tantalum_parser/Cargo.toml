[package]
name = "tantalum_parser"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_ast = { path = "../tantalum_ast" }
tantalum_lexer = { path = "../tantalum_lexer" }
tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}

[lints]
workspace = true
