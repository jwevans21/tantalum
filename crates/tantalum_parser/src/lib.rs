use tantalum_lexer::Lexer;
use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxNode, SyntaxToken, SyntaxTree};

#[cfg(test)]
mod tests;

mod expressions;
mod identifiers;
mod items;
mod paths;
mod statements;
mod types;
mod whitespace;

/// The different events that can be emitted when parsing.
///
/// These will determine if a new tree should be opened, closed, or a token should be added to the
/// currently opened tree.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
enum ParserEvent {
    /// Open a new token of the specified [`SyntaxKind`].
    ///
    /// This will be resolved into a [`SyntaxTree`] when closed.
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    /// [`SyntaxTree`]: tantalum_syntax::SyntaxTree
    Open {
        kind: SyntaxKind,
        start: SourceLocation,
    },
    /// Close the current open [`SyntaxTree`].
    ///
    /// [`SyntaxTree`]: tantalum_syntax::SyntaxTree
    Close { end: SourceLocation },
    /// Add a terminal token to the current open [`SyntaxTree`].
    ///
    /// This will be transformed into a [`SyntaxNode::Token`] when the tree is constructed.
    ///
    /// [`SyntaxTree`]: tantalum_syntax::SyntaxTree
    /// [`SyntaxNode::Token`]: tantalum_syntax::SyntaxNode::Token
    Token { kind: SyntaxKind, span: SourceSpan },
}

/// A mark for an event that has been opened. Used by the parser to then either close or adjust the
/// currently opened tree.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(transparent)]
struct MarkOpened {
    index: usize,
}

/// A mark for an event that has been closed. Used to update the tree structure based on later
/// tokens.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(transparent)]
struct MarkClosed {
    open_index: usize,
}

/// The parser for the Tantalum language.
///
/// Controls the parsing by emitting events that are then used to construct the syntax tree.
///
/// All parsing functions are implemented as functions that take a `&mut Parser`. The exposed
/// interface are the [`Parser::parse`] and [`Parser::finish`] functions.
///
/// The interface of the parser is based on [Resilient LL Parsing Tutorial] which is licence under
/// [MIT OR Apache-2.0].
///
/// [Resilient LL Parsing Tutorial]: https://matklad.github.io/2023/05/21/resilient-ll-parsing-tutorial.html
/// [MIT OR Apache-2.0]: https://matklad.github.io/about.html
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Parser<'file> {
    /// The lexer to use for parsing.
    lexer: Lexer<'file>,
    /// The source file being parsed.
    source: &'file SourceFile,
    /// The events emitted by the parser.
    events: Vec<ParserEvent>,
}

impl<'file> Parser<'file> {
    /// Create a new [`Parser`] from a [`Lexer`].
    ///
    /// [`Lexer`]: tantalum_lexer::Lexer
    #[must_use]
    pub fn new(lexer: Lexer<'file>) -> Self {
        let source = lexer.source();
        Self {
            lexer,
            source,
            events: Vec::new(),
        }
    }

    /// Get a string of the specified `SyntaxToken` from the source file.
    #[must_use]
    pub fn get_span_str(&self, span: &SourceSpan) -> Option<&str> {
        self.source.read_span(*span)
    }

    /// Parse the source file.
    ///
    /// This just runs the `file` parser.
    ///
    /// To get the resulting `SyntaxTree`, call `finish`.
    ///
    /// ```rust
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_lexer::Lexer;
    /// use tantalum_parser::Parser;
    ///
    /// let mut files = SourceFileCollection::new();
    /// let file_id = files.add_file("main.tan".to_string(), "fn main() {}".to_string());
    /// let file = files.get_file(file_id).expect("should have file, since it was just added");
    ///
    /// let mut parser = Parser::new(Lexer::new(file));
    ///
    /// parser.parse();
    ///
    /// let tree = parser.finish();
    /// ```
    pub fn parse(&mut self) {
        use crate::items::file;
        let _ = file(self);
    }

    /// Finish the parser and return the resulting `SyntaxTree`.
    ///
    /// # Panics
    ///
    /// - if there are unmatched open events.
    /// - if there are unmatched close events.
    /// - if there are unexpected token events (a token must be inside a tree).
    ///
    /// ```rust
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_lexer::Lexer;
    /// use tantalum_parser::Parser;
    ///
    /// let mut files = SourceFileCollection::new();
    /// let file_id = files.add_file("main.tan".to_string(), "fn main() {}".to_string());
    /// let file = files.get_file(file_id).expect("should have file, since it was just added");
    ///
    /// let mut parser = Parser::new(Lexer::new(file));
    ///
    /// parser.parse();
    ///
    /// let tree = parser.finish();
    /// ```
    #[must_use]
    pub fn finish(self) -> SyntaxTree {
        let mut stack = Vec::new();

        for event in self.events {
            match event {
                ParserEvent::Open { kind, start } => {
                    stack.push(SyntaxTree::new(
                        kind,
                        SourceSpan::new(start.file(), start.offset(), start.offset()),
                    ));
                }
                ParserEvent::Close { end } => {
                    let mut tree = stack.pop().expect("unmatched close event");
                    tree.extend_span(end);

                    if let Some(parent) = stack.last_mut() {
                        parent.add_child(SyntaxNode::Tree(tree));
                    } else {
                        return tree;
                    }
                }
                ParserEvent::Token { kind, span } => {
                    let token = SyntaxToken::new(kind, span);
                    if let Some(parent) = stack.last_mut() {
                        parent.add_child(SyntaxNode::Token(token));
                    } else {
                        panic!("unexpected token event, a token must be inside a tree");
                    }
                }
            }
        }

        assert_eq!(stack.len(), 1, "unmatched open event");
        stack.pop().expect("unmatched open event")
    }

    /// Open a new token of the specified [`SyntaxKind`].
    ///
    /// This resolves into a [`SyntaxTree`] when closed using [`Parser::close`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    /// [`SyntaxTree`]: tantalum_syntax::SyntaxTree
    /// [`Parser::close`]: Parser::close
    #[must_use]
    pub(crate) fn open(&mut self, kind: SyntaxKind) -> MarkOpened {
        let start = self.lexer.location();
        self.events.push(ParserEvent::Open { kind, start });
        MarkOpened {
            index: self.events.len() - 1,
        }
    }

    /// Open a new token of the specified [`SyntaxKind`] before the specified mark.
    ///
    /// Useful for when adjusting the tree based on later tokens.
    ///
    /// Similarly to [`Parser::open`], this resolves into a [`SyntaxTree`] when closed using [`Parser::close`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    /// [`SyntaxTree`]: tantalum_syntax::SyntaxTree
    /// [`Parser::open`]: Parser::open
    /// [`Parser::close`]: Parser::close
    pub(crate) fn open_before(&mut self, close: MarkClosed, kind: SyntaxKind) -> MarkOpened {
        match self.events.get(close.open_index) {
            Some(ParserEvent::Open { start, .. }) => {
                self.events.insert(
                    close.open_index,
                    ParserEvent::Open {
                        kind,
                        start: *start,
                    },
                );
                MarkOpened {
                    index: close.open_index,
                }
            }
            _ => panic!("can only open before other open events"),
        }
    }

    /// Adjust the kind of the current open token.
    ///
    /// # Panics
    ///
    /// - if the mark is not an open event.
    pub(crate) fn adjust(&mut self, mark: MarkOpened, updated_kind: SyntaxKind) {
        if let Some(ParserEvent::Open { kind, .. }) = self.events.get_mut(mark.index) {
            *kind = updated_kind;
        } else {
            panic!("can only adjust open events");
        }
    }

    /// Add a token to the current open tree.
    ///
    /// When the tree is constructed the token will be transformed into a `SyntaxNode::Token`.
    pub(crate) fn token(&mut self, token: SyntaxToken) -> MarkClosed {
        self.events.push(ParserEvent::Token {
            kind: token.kind(),
            span: token.span(),
        });

        MarkClosed {
            open_index: self.events.len() - 1,
        }
    }

    /// Close the current open token.
    #[must_use]
    pub(crate) fn close(&mut self, mark: MarkOpened) -> MarkClosed {
        let end = self.lexer.location();
        self.events.push(ParserEvent::Close { end });
        MarkClosed {
            open_index: mark.index,
        }
    }

    /// Error with the specified message.
    pub(crate) fn error(&mut self, mark: MarkOpened, message: &str) -> MarkClosed {
        eprintln!("error: {message}");
        self.adjust(mark, SyntaxKind::Error);
        self.close(mark)
    }

    /// Peek the `n` tokens ahead.
    ///
    /// Does not consume the tokens or adjust the state of the lexer.
    #[must_use]
    pub(crate) fn peek(&self, n: usize) -> Option<SyntaxToken> {
        self.lexer.peek_token(n)
    }

    /// Consume the current [`SyntaxToken`], if it exists add it to the tree structure as
    /// a [`SyntaxNode::Token`].
    ///
    /// Returns the consumed token if it exists.
    ///
    /// [`SyntaxToken`]: tantalum_syntax::SyntaxToken
    /// [`SyntaxNode::Token`]: tantalum_syntax::SyntaxNode::Token
    #[must_use]
    pub(crate) fn consume(&mut self) -> Option<SyntaxToken> {
        let token = self.lexer.next_token()?;

        self.token(token);

        Some(token)
    }

    /// Advance the lexer and return the next token.
    ///
    /// This does not add the token to the tree structure.
    pub(crate) fn advance(&mut self) -> Option<SyntaxToken> {
        self.lexer.next_token()
    }

    /// Check if the parser is at the end of the source file.
    pub(crate) fn at_end(&self) -> bool {
        self.peek(1).is_none()
    }

    /// Check if the parser is at the specified [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    #[must_use]
    pub(crate) fn is_at(&self, kind: SyntaxKind) -> bool {
        self.peek(1).is_some_and(|token| token.kind() == kind)
    }

    /// Check if the parser is at any of the specified [`SyntaxKind`]s.
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    #[must_use]
    pub(crate) fn is_at_any(&self, kinds: &[SyntaxKind]) -> bool {
        self.peek(1)
            .is_some_and(|token| kinds.contains(&token.kind()))
    }

    /// Check if the parser is at the specified [`SyntaxKind`] and consume it if it is.
    ///
    /// Returns the consumed token and adds it to the tree structure if it exists.
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    #[must_use]
    pub(crate) fn consume_if(&mut self, kind: SyntaxKind) -> Option<SyntaxToken> {
        if self.is_at(kind) {
            self.consume()
        } else {
            None
        }
    }

    /// Check if the parser is at any of the specified [`SyntaxKind`]s and consume it if it is.
    ///
    /// Returns the consumed token and adds it to the tree structure if it exists.
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    #[must_use]
    pub(crate) fn consume_if_any(&mut self, kinds: &[SyntaxKind]) -> Option<SyntaxToken> {
        if self.is_at_any(kinds) {
            self.consume()
        } else {
            None
        }
    }

    /// Collect the tokens into a `Vec` while the specified function returns `true`.
    ///
    /// Advances the parser, but does not add the tokens to the tree structure.
    pub(crate) fn advance_while<F>(&mut self, mut f: F) -> Vec<SyntaxToken>
    where
        F: FnMut(SyntaxKind) -> bool,
    {
        let mut tokens = Vec::new();

        while let Some(token) = self.peek(1) {
            if f(token.kind()) {
                tokens.push(self.advance().unwrap());
            } else {
                break;
            }
        }

        tokens
    }
}
