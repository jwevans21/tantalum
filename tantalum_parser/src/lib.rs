use error::ParseError;
use tantalum_ast::TopLevelExpression;
use tantalum_lexer::{token::Token, token_kind::TokenKind, Lexer};
use tantalum_span::Span;

pub mod error;

mod expressions;
mod statements;
mod types;
mod top_level;

#[cfg(test)]
mod tests;

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct Parser<'file_name, 'source> {
    // lexer: Lexer<'file_name, 'source>,
    source: &'source str,
    file_name: &'file_name str,
    tokens: Vec<Token<'file_name, 'source>>,
    eof: Span<'file_name>,
    position: usize,
}

impl<'file_name, 'source> Parser<'file_name, 'source> {
    #[must_use]
    #[inline]
    pub fn new(mut lexer: Lexer<'file_name, 'source>) -> Self {
        Self {
            source: lexer.source(),
            file_name: lexer.file_name(),
            tokens: lexer.by_ref().collect(),
            eof: lexer.span(),
            position: 0,
        }
    }

    /// Parse the entire source file.
    /// 
    /// # Errors
    /// 
    /// Returns an error if the parser encounters an unexpected token or the end of the file.
    pub fn parse(
        &mut self,
    ) -> Result<Vec<TopLevelExpression<'file_name, 'source>>, error::ParseError<'file_name, 'source>>
    {
        let mut top_levels = Vec::new();

        while !self.is_eof() {
            top_levels.push(self.parse_top_level()?);
        }

        Ok(top_levels)
    }

    fn is_eof(&self) -> bool {
        self.position >= self.tokens.len()
    }

    fn is_at(&self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position)
            .filter(|token| token.kind() == kind)
            .copied()
    }

    fn advance_if(&mut self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        if let Some(token) = self.is_at(kind) {
            self.position += 1;
            Some(token)
        } else {
            None
        }
    }

    fn expect(
        &mut self,
        kind: TokenKind,
    ) -> Result<Token<'file_name, 'source>, error::ParseError<'file_name, 'source>> {
        let Some(token) = self.tokens.get(self.position) else {
            return Err(error::ParseError::unexpected_eof(self.source, self.eof));
        };

        if token.kind() == kind {
            self.position += 1;
            Ok(*token)
        } else {
            Err(error::ParseError::unexpected_token_set(
                self.source,
                token.span(),
                token.kind(),
                &[kind],
            ))
        }
    }

    #[allow(unused)]
    fn next_is(&self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position + 1)
            .filter(|token| token.kind() == kind)
            .copied()
    }

    fn is_at_any<'a>(&self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position)
            .filter(|token| set.contains(&token.kind()))
            .copied()
    }

    fn advance_if_any<'a>(&mut self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        if let Some(token) = self.is_at_any(set) {
            self.position += 1;
            Some(token)
        } else {
            None
        }
    }

    fn expect_any<'a>(
        &mut self,
        set: &'a [TokenKind],
    ) -> Result<Token<'file_name, 'source>, error::ParseError<'file_name, 'source>> {
        let Some(token) = self.tokens.get(self.position) else {
            return Err(error::ParseError::unexpected_eof(self.source, self.eof));
        };

        if set.contains(&token.kind()) {
            self.position += 1;
            Ok(*token)
        } else {
            Err(error::ParseError::unexpected_token_set(
                self.source,
                token.span(),
                token.kind(),
                set,
            ))
        }
    }

    #[allow(unused)]
    fn next_is_any<'a>(&self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position + 1)
            .filter(|token| set.contains(&token.kind()))
            .copied()
    }

    fn nth(&self, n: usize) -> Option<Token<'file_name, 'source>> {
        self.tokens.get(self.position + n).copied()
    }

    fn put_back(&mut self, n: usize) {
        self.position -= n;
    }
}
