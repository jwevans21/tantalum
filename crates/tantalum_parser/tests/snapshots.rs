use insta::assert_ron_snapshot;

use tantalum_lexer::Lexer;
use tantalum_parser::Parser;

macro_rules! snapshot_from_file {
    ($file_name:ident) => {
        #[test]
        fn $file_name() {
            let source = include_str!(concat!("../../examples/", stringify!($file_name), ".ta"));
            let lexer = Lexer::new(concat!(stringify!($file_name), ".ta"), source);

            let mut parser = Parser::new(lexer);

            let ast = parser.parse();

            assert_ron_snapshot!(ast);
        }
    };
    ($($file_name:ident),*) => {
        $(snapshot_from_file! { $file_name })*
    };
}

snapshot_from_file! {
    start,
    hello_world,
    conditionals,
    loops
}
