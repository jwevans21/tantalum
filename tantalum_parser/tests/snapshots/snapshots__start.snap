---
source: tantalum_parser/tests/snapshots.rs
expression: ast
snapshot_kind: text
---
Ok([
  TopLevelExpression(
    span: Span(
      start: Location(
        file_name: "start.ta",
        position: 0,
        line: 1,
        column: 1,
      ),
      end: Location(
        file_name: "start.ta",
        position: 32,
        line: 3,
        column: 2,
      ),
    ),
    kind: FunctionDeclaration(
      name: "main",
      parameters: [],
      return_type: Type(
        span: Span(
          start: Location(
            file_name: "start.ta",
            position: 11,
            line: 1,
            column: 12,
          ),
          end: Location(
            file_name: "start.ta",
            position: 14,
            line: 1,
            column: 15,
          ),
        ),
        kind: Named("i32"),
      ),
      body: Statement(
        span: Span(
          start: Location(
            file_name: "start.ta",
            position: 15,
            line: 1,
            column: 16,
          ),
          end: Location(
            file_name: "start.ta",
            position: 32,
            line: 3,
            column: 2,
          ),
        ),
        kind: Block(
          statements: [
            Statement(
              span: Span(
                start: Location(
                  file_name: "start.ta",
                  position: 21,
                  line: 2,
                  column: 5,
                ),
                end: Location(
                  file_name: "start.ta",
                  position: 30,
                  line: 2,
                  column: 14,
                ),
              ),
              kind: Return(
                value: Some(Expression(
                  span: Span(
                    start: Location(
                      file_name: "start.ta",
                      position: 28,
                      line: 2,
                      column: 12,
                    ),
                    end: Location(
                      file_name: "start.ta",
                      position: 29,
                      line: 2,
                      column: 13,
                    ),
                  ),
                  kind: IntegerLiteral(
                    value: "0",
                    radix: 10,
                  ),
                )),
              ),
            ),
          ],
        ),
      ),
    ),
  ),
])
